### PR TITLE
WT-9495 Create debug flag to make checkpoint perform eviction on reconcilling pages

### PR DESCRIPTION
**Pull request by jiechenbo**
This pull request deals with adding a debug flag to force checkpoint to perform evict when reconciling pages. I have also added the config reproducer, that reproduces the error seen in WT-9477.

### CODE REVIEW

**Review by etienneptl — APPROVED** [review]
Awesome work on the config, thanks @jiechenbo, LGTM!

File: dist/api_data.py [hunk 1]
```diff
@@ -560,6 +560,9 @@ def __ge__(self, other):
         Config('corruption_abort', 'true', r'''
             if true and built in diagnostic mode, dump core in the case of data corruption''',
             type='boolean'),
+        Config('checkpoint_evict_page', 'false', r'''
+            if true, when performing checkpoint evict all pages that are going to be reconciled.''',
```
Comment (kommiharibabu) [inline dist/api_data.py:564]: all `leaf` pages?

File: test/suite/test_debug_mode10.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    conn_config = 'cache_size=1000MB,statistics=(all),eviction_dirty_target=80,eviction_dirty_trigger=95,debug_mode=(checkpoint_evict_page=true)'
+    uri = "table:test_debug_mode10"
+
+    def test_checkpoint_evict_page(self):
+        self.session.create(self.uri, 'key_format=i,value_format=S')
+
+        cursor = self.session.open_cursor(self.uri)
+        for i in range(0, 10000):
+            self.session.begin_transaction()
+            cursor[i] = 'b' * 5000
```
Comment (kommiharibabu) [inline test/suite/test_debug_mode10.py:47]: Do we really need such a big value to reproduce this issue?

**Review by kommiharibabu — APPROVED** [review]
LGTM.

**Comment by quchenhao** [comment]
@jiechenbo @etienneptl  Instead of a debug mode, I'd prefer this being implemented as a timing stress. Because this is not something we tune when debugging I think. It more fits as a fail point, which is implemented as a timing stress for now.

**Comment by etienneptl** [comment]
> @jiechenbo @etienneptl Instead of a debug mode, I'd prefer this being implemented as a timing stress. Because this is not something we tune when debugging I think. It more fits as a fail point, which is implemented as a timing stress for now.

This is a good idea @quchenhao and it makes sense. As we need this to unblock the 6.0 release I would keep it as is and create a ticket to improve this if it's too much work.

**Review by etienneptl — APPROVED** [review]

File: src/btree/bt_sync.c [hunk 3]
```diff
@@ -607,7 +607,9 @@ __wt_sync_file(WT_SESSION_IMPL *session, WT_CACHE_OP syncop)
              * Once the transaction has given up it's snapshot it is no longer safe to reconcile
              * pages. That happens prior to the final metadata checkpoint.
              */
-            if (F_ISSET(walk, WT_REF_FLAG_LEAF) && page->read_gen == WT_READGEN_WONT_NEED &&
+            if (F_ISSET(walk, WT_REF_FLAG_LEAF) &&
+              (page->read_gen == WT_READGEN_WONT_NEED ||
+                FLD_ISSET(conn->timing_stress_flags, WT_TIMING_STRESS_CHECKPOINT_EVICT_PAGE)) &&
```
Comment (agorrod) [inline src/btree/bt_sync.c:612]: Is it worth making this more subtle - and evicting a random proportion of pages, rather than every page?

Comment (jiechenbo) [reply src/btree/bt_sync.c:612]: I agree, although I would prefer merging this in with a WT reproducer first, and then create a follow-up ticket on deciding on any further improvements regarding this flag. If we evict a random portion of pages the chances of hitting the bug will be alot lower and even possibly go away.

File: test/suite/test_debug_mode10.py (hunk 2 above)
Comment (jiechenbo) [reply test/suite/test_debug_mode10.py]: The python test doesn't reproduce the issue. The python test is to just make sure that we are evicting pages during checkpoint. So the big value is just to produce more leaf pages.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Comment by kommiharibabu** [comment]
@jiechenbo Thanks for fixing the test/format timing stress problem.

File: src/btree/bt_sync.c (hunk 3 above)
Comment (agorrod) [reply src/btree/bt_sync.c:612]: sgtm

**Comment by jiechenbo** [comment]
evergreen merge

**Pull request merged**